src/main/java/io/vertx/core/http/impl/VertxHttp2ConnectionHandler.java

```diff
@@ -205,17 +205,37 @@ private void _writeHeaders(Http2Stream stream, Http2Headers headers, boolean end
 
   void writeData(Http2Stream stream, ByteBuf chunk, boolean end) {
     EventExecutor executor = chctx.executor();
+    ChannelPromise promise = chctx.newPromise();
+    if (executor.inEventLoop()) {
+      _writeData(stream, chunk, end, promise);
+    } else {
+      executor.execute(() -> {
+        _writeData(stream, chunk, end, promise);
+      });
+    }
+  }
+
+  void writeData(Http2Stream stream, ByteBuf chunk, boolean end, Handler<AsyncResult<Void>> handler) {
```

Comment:
you can keep a single `writeData` method and add the handler as listener of the promise when it's not null and remove `void writeData(Http2Stream stream, ByteBuf chunk, boolean end)`